display_list/display_list_builder.cc

```diff
@@ -1046,10 +1216,180 @@ void DisplayListBuilder::drawShadow(const SkPath& path,
+
   transparent_occluder  //
       ? Push<DrawShadowTransparentOccluderOp>(0, 1, path, color, elevation, dpr)
       : Push<DrawShadowOp>(0, 1, path, color, elevation, dpr);
   UpdateLayerOpacityCompatibility(false);
 }
 
+bool DisplayListBuilder::ComputeFilteredBounds(SkRect& bounds,
+                                               const DlImageFilter* filter) {
+  if (filter) {
+    if (!filter->map_local_bounds(bounds, bounds)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool DisplayListBuilder::AdjustBoundsForPaint(SkRect& bounds,
+                                              DisplayListAttributeFlags flags) {
+  if (flags.ignores_paint()) {
+    return true;
+  }
+
+  if (flags.is_geometric()) {
+    // Path effect occurs before stroking...
+    DisplayListSpecialGeometryFlags special_flags =
+        flags.WithPathEffect(current_.getPathEffectPtr());
+    if (current_.getPathEffect()) {
+      auto effect_bounds = current_.getPathEffect()->effect_bounds(bounds);
+      if (!effect_bounds.has_value()) {
+        return false;
+      }
+      bounds = effect_bounds.value();
+    }
+
+    if (flags.is_stroked(current_.getDrawStyle())) {
+      // Determine the max multiplier to the stroke width first.
+      SkScalar pad = 1.0f;
+      if (current_.getStrokeJoin() == DlStrokeJoin::kMiter &&
+          special_flags.may_have_acute_joins()) {
+        pad = std::max(pad, miter_limit_);
+      }
+      if (current_.getStrokeCap() == DlStrokeCap::kSquare &&
+          special_flags.may_have_diagonal_caps()) {
+        pad = std::max(pad, SK_ScalarSqrt2);
+      }
+      pad *= half_stroke_width_;
+      bounds.outset(pad, pad);
+    }
+  }
+
+  if (flags.applies_mask_filter()) {
+    if (current_.getMaskFilter()) {
+      const DlBlurMaskFilter* blur_filter = current_.getMaskFilter()->asBlur();
+      if (blur_filter) {
+        SkScalar mask_sigma_pad = blur_filter->sigma() * 3.0;
+        bounds.outset(mask_sigma_pad, mask_sigma_pad);
+      } else {
+        SkPaint p;
+        p.setMaskFilter(current_.getMaskFilter()->skia_object());
+        if (!p.canComputeFastBounds()) {
+          return false;
+        }
+        bounds = p.computeFastBounds(bounds, &bounds);
+      }
+    }
+  }
+
+  if (flags.applies_image_filter()) {
+    return ComputeFilteredBounds(bounds, current_.getImageFilter().get());
+  }
+
+  return true;
+}
+
+bool DisplayListBuilder::AccumulateUnbounded() {
+  accumulator()->accumulate(current_layer_->clip_bounds);
+  return true;
+}
+
+bool DisplayListBuilder::AccumulateOpBounds(SkRect& bounds,
+                                            DisplayListAttributeFlags flags) {
+  if (AdjustBoundsForPaint(bounds, flags)) {
+    return AccumulateBounds(bounds);
+  } else {
+    return AccumulateUnbounded();
+  }
+}
+bool DisplayListBuilder::AccumulateBounds(SkRect& bounds) {
+  getTransform().mapRect(&bounds);
+  if (bounds.intersect(current_layer_->clip_bounds)) {
```

Comment:
Where is the call to has_clip() here? It was in the code that was copied. If there is no clip then this intersection will fail.